### PR TITLE
Install script: Windows-compatible cleanup path.

### PR DESCRIPTION
**Pull request by criemen**
Windows might need some retrying around deleting
the target directory.

### CODE REVIEW

File: misc/bazel/internal/install.py [hunk 1]
```diff
@@ -42,8 +44,20 @@
 script = runfiles.Rlocation(opts.pkg_install_script)
 
 if destdir.exists() and opts.cleanup:
-    shutil.rmtree(destdir)
-
+    if platform.system() == 'Windows':
+        # On Windows we might have virus scanner still looking at the path so
+        # attempt removal a couple of times sleeping between each attempt.
+        for attempt in [1, 2]:
```
Comment (redsun82) [inline misc/bazel/internal/install.py:50]: I think there's a bug here:
1. we enter the loop with `attempt == 1`.
1. we try `rmtree`. Let's suppose we fail with `winerror == 32`
2. we sleep 1s
3. we reenter the loop with `attempt == 2`.
4. we try `rmtree` again. Let's suppose we fail again with `winerror == 32`
5. we sleep for 2s, but then...
6. ...we exit the loop and we move on, having silently failed to remove the directory, and having slept 2s for no reason

we need to do a final `rmtree` after the `for` unless we succeeded. We could do that by adding a
```python
else:
   shutil.rmtree(destdir)
```
after the `for` loop, but I know not all are comfortable with the `for: ... else: ...` construct (`else` is visited only if the `for` was not interrupted by a `break`)

File: misc/bazel/internal/install.py [hunk 2]
```diff
@@ -42,8 +44,20 @@
...
+            try:
+                shutil.rmtree(destdir)
+                break
+            except OSError as e:
+                if e.winerror == 32:
```
Comment (redsun82) [inline misc/bazel/internal/install.py:55]: can we give a name for this 32 magic number?

**Review by redsun82 — CHANGES_REQUESTED** [review]
I think there's a bug in the current code, or am I missing something?

File: misc/bazel/internal/install.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (criemen) [reply misc/bazel/internal/install.py:50]: it was 1, 2, 2 when I introduced it in the internal pytest codebase, but the original code we're currently using is 1, 2.

I can add the second 2 back, though.

File: misc/bazel/internal/install.py (hunk 2 above)
Comment (criemen) [reply misc/bazel/internal/install.py:55]: I don't think python exposes a public api for this, they themselves do this: https://github.com/python/cpython/blob/8ad6067bd4556afddc86004f8e350aa672fda217/Lib/shutil.py#L456

File: misc/bazel/internal/install.py (hunk 1 above)
Comment (criemen) [reply misc/bazel/internal/install.py]: Fixed. The code is _exceedingly_ ugly now though :(

[the remark this answers is not in the document]
Comment (redsun82) [reply misc/bazel/internal/install.py]: I don't have any particular preference on this, I'm ok with this being `[1, 2]`

File: misc/bazel/internal/install.py (hunk 2 above)
Comment (redsun82) [reply misc/bazel/internal/install.py:55]: yeah, I found out as much, but we can still give it a name in this script to help the reader :slightly_smiling_face:

Comment (criemen) [reply misc/bazel/internal/install.py:55]: done

Comment (redsun82) [reply misc/bazel/internal/install.py]: actually, as far as I could test an `OSError` with `.winerror == 32` actually gets raised as a more specific `PermissionError`, but I guess the `.winerror == 32` specifies this is the "still in use" error. So I just suggest we add a `still_in_use_winerror = 32` and use that for readability

File: misc/bazel/internal/install.py [hunk 3]
```diff
@@ -42,7 +44,22 @@
...
+                    time.sleep(retry_delay)
+                else:
+                    raise
+        else:
+            shutil.rmtree(destdir)
+    else:
+        shutil.rmtree(destdir)
```
Comment (redsun82) [inline misc/bazel/internal/install.py:64]: maybe to lower the ugliness we can extract our own `rmtree`, replace `break` with return, and collapse these two `else`s into one fall-through 
```suggestion
    shutil.rmtree(destdir)
```
?

**Pull request merged**